### ISSUE TITLE
[Raid][NPC][Lady Vash'j] Persuasion again and general MC AI behaviour

### ISSUE THREAD

**Issue by Brasso** [Atlantiss/NetherwingBugtracker#3541]
[//]: # (Enclose links to things related to the bug using http://wowhead.com or any other TBC database.)
[//]: # (You can use screenshot ingame to visual the issue.)
[//]: # (Write your tickets according to the format:)
[//]: # ([Quest][Azuremyst Isle] Red Snapper - Very Tasty!)
[//]: # ([NPC] Magistrix Erona)
[//]: # ([Spell][Mage] Fireball)
[//]: # ([Npc][Drop] Ghostclaw Lynx)
[//]: # ([Web] Armory doesnt work)
[//]: # (To find server revision type `.server info` command in the game chat.)

After discussion on discord I've decided to reopen the MC issue thread and do some more investigation.

### Obvious bugs:
Fixed
~~https://clips.twitch.tv/SweetTangentialKiwiBIRB - Pet targetted by MC~~
~~https://clips.twitch.tv/BlueEagerSardineAMPTropPunch - Elemental targetted by MC~~

~~https://clips.twitch.tv/ScaryBraveSharkNomNom - MC'd player kills another player in the MC ends, leaving only 1 MC target up. Should be 2 targets at all times.~~
~~https://clips.twitch.tv/GeniusAthleticBulgogiVoHiYo - MC ends prematurely again for the warrior, looks like someone died.~~

### MC and other CC effects sometimes teleport targets to the position the player had when CC began:
https://clips.twitch.tv/DreamyCaringMouseUnSane - MC teleport again.
https://imgur.com/a/JOVrSjs - MC lands on player in the first pic, the second pic shows the position where he gets teleported to, which is the initial position the CC landed on.

https://clips.twitch.tv/AmazonianBlightedHerbsNomNom - Tank teleport.

Other threads mentioning this:
#1956 #3327 

### NW Behaviour - MC AI:

Now to the part the requires some research, after watching the test stream of Vash'j and comparing it to the Method kill video. I do believe melee is functioning correctly on NW. However, it seems that healers and casters do not. From what I've observed, albeit limited footage due to only one POV and MC'd targets being CC'd, healers and casters do not use their spells and heals as much as they should.

**Healers:**
https://clips.twitch.tv/MildRelatedClamJonCarnage - Notice the paladin doesn't cast a single spell. Healer.
https://clips.twitch.tv/FancyAssiduousRuffUWot - Shaman teleports, and uses no spells? Healer again.

**Casters:**
The main thing that should be observed in the following clips is the amount of meleeing and the downtime between casters using their spells. In these clips the casters use on average 2 spells.

https://clips.twitch.tv/AuspiciousDirtyHerdJonCarnage 
https://clips.twitch.tv/FrigidAmazingSwallowEleGiggle 

### Retail TBC Behaviour:

https://www.youtube.com/watch?v=SYypbzs9d44&feature=youtu.be&t=551 - Notice how the shaman healer immediately casts an earth shock, followed by chain lightning and then lightning bolt. After a couple of seconds she starts using heals on Vash'j, 1 goes through the other doesn't have time to finish. 5 spells in rapid succession.

https://youtu.be/SYypbzs9d44?t=688 - The druid uses 3 spells successfully, the 4th doesn't have time to finish.

https://youtu.be/SYypbzs9d44?t=744 - Correct me if I'm seeing wrong here, but the paladin followed targets, but never melees them. Uses 5 spells in total, 3 of which are heals/buffs on Vash'j.

https://www.youtube.com/watch?v=SYypbzs9d44&feature=youtu.be&t=707 - Hard to see, but I don't think the gnome melees once.

https://youtu.be/SYypbzs9d44?t=553 - Notice the Gnome mage, uses in total 3 spells, 1 cast, 2 channeled throughout the entire duration, no pause in between, no melee.

https://youtu.be/SYypbzs9d44?t=640 - The warlock starts casting a spell immediately after being MC'd, spell goes through (earth shock is too late), instantly starts channeling rain of fire. She then gets successfully interrupted and immediately starts casting from the shadow school instead. No melee.

### Kael World First MC Behaviour:
Since this is probably MC AI in general I will add retail footage from Kael world first kill aswell.

https://youtu.be/QMNWn5RAays?t=140 - Mage and shadow priest uses spells the entire duration of their MC, no melee. There is some downtime for the shadow priest between casts, but she never draws weapon.

https://youtu.be/QMNWn5RAays?t=164 - The shaman continuously uses spells until he is CC'd. **First footage of a caster using melee here**, seems to be possible but rare.

https://youtu.be/QMNWn5RAays?t=205 - No melee from casters, some downtime but casts spells regularly.

### Conclusion:
The AI for Persuasion on NW is wrong. There should be almost no downtime between spell casts for healers and casters. Casters should ***almost*** never melee their targets and if they get interrupted they should immediately switch an available school as start casting that. Healers should use buffs and heals on Vash'j, as well as using offensive spells on other players.

This mechanic is vital to the encounter and I believe it is very important to get right. In the current state, the boss is much easier than it should be.

### List of spells used:
<details open>
<summary>Shaman</summary>
<br>

- Water shield
- Earth shield
- Lightning shield
- Lightning bolt
- Chain lightning
- Earth shock
- Heroism
- Healing wave
- Chain Heal

</details>
<details open>
<summary>Druid</summary>
<br>

- Cyclone
- Starfire
- Hurricane
- Wrath
- Healing Touch
- All hots

</details>
<details open>
<summary>Paladin</summary>
<br>

- All Auras
- Lots of seals, maybe all
- Blessings, hard to tell which are used but BoP is not in this list
- Holy shock
- Flash of Light
- Holy light
- Consecration

</details>
<details open>
<summary>Rogue</summary>
<br>

- Throw
- Blade flurry
- Evasion
- Adrenaline rush
- Kick

</details>
<details open>
<summary>Priest</summary>
<br>

- Smite
- Flash Heal
- Inner fire
- Shadow word: Pain
- Mind Flay
- Vampiric Embrace

</details>
<details open>
<summary>Mage</summary>
<br>

- Blizzard
- Evocation
- Ice lance
- Fireball
- Fireblast
- Flamestrike

</details>
<details open>
<summary>Warlock</summary>
<br>

- Curse of agony
- Curse of tongues
- Seed of Corruption
- Immolate
- Shadow bolt
- Rain of Fire
- Incinerate

</details>
<details open>
<summary>Warrior</summary>
<br>

- Battleshout
- Berserker Rage
- Blood Rage
- Cleave

</details>
<details open>
<summary>Hunter</summary>
<br>

- All aspects
- Volley
- Arcane shot

</details>

Reference videos:
https://www.youtube.com/watch?v=SYypbzs9d44 - Vash'j world second
https://www.youtube.com/watch?v=QMNWn5RAays - Kael world first
https://www.youtube.com/watch?v=8r6pCxK9hDE
https://www.youtube.com/watch?v=9rRTu6QmxFg

#3542 Another thread listing spells used by MC AI

**Server Revision**: 
Pre rev 2434 (only CC was fixed after this)

**Comment by SwiftieTBC** [comment]
After looking through the clips and checking against quite a lot of my research. I would agree that the chance to roll a melee should be reduced for ranged dps and healers. With that being said it is important to observe that the majority of the ranged DPS spend far more time AOEing random spots on the ground and Healers spend far more time doing non healing abilities than healing abilities. And in neither of these cases does it seem the aim of the AI is to maximize DPS or Healing.

**Comment by Brasso** [comment]
I agree Swiftie, the aim of the AI is NOT to maximize DPS or healing. The spell usage seems to be completely random. But using a spell has a much, much higher priority than following players or meleeing. Healers seems to prioritize damage spells over healing spells, but both come before melee.

**Comment by Tenoutoften** [comment]
Are MCed players by different mobs, using different MC spells using the same spells?

**Comment by Easyfraggy** [comment]
I think this was already changed over a week ago @Wolffenstein ?

**Comment by Godmatik** [comment]
Rev 4018

Mind Controls from AI (not just Vashj) still needs a full rescript/revamp.

Right now many abilities are not present, for example it is impossible for a mind controlled player to heal Vashj when MCed

**Comment by Godmatik** [comment]
Player AI for mind controls have been fully revamped
